doc/source/reference/c-api/array.rst

```diff
@@ -2353,59 +2353,61 @@ it is possible to do this.
-        /* Free the memory owned by this auxdata */
-        NPY_AUXDATA_FREE(d->funcdata);
-        PyArray_free(d);
-    }
-
-    NpyAuxData *clone_element_doubler_aux_data(NpyAuxData *data)
-    {
-        eldoubler_aux_data *ret = PyArray_malloc(sizeof(eldoubler_aux_data));
-        if (ret == NULL) {
-            return NULL;
-        }
-
-        /* Raw copy of all data */
-        memcpy(ret, data, sizeof(eldoubler_aux_data));
-
-        /* Fix up the owned auxdata so we have our own copy */
-        ret->funcdata = NPY_AUXDATA_CLONE(ret->funcdata);
-        if (ret->funcdata == NULL) {
-            PyArray_free(ret);
-            return NULL;
-        }
-
-        return (NpyAuxData *)ret;
-    }
-
-    NpyAuxData *create_element_doubler_aux_data(
-                                ElementCopier_Func *func,
-                                NpyAuxData *funcdata)
-    {
-        eldoubler_aux_data *ret = PyArray_malloc(sizeof(eldoubler_aux_data));
-        if (ret == NULL) {
-            PyErr_NoMemory();
-            return NULL;
-        }
-        memset(&ret, 0, sizeof(eldoubler_aux_data));
-        ret->base->free = &free_element_doubler_aux_data;
-        ret->base->clone = &clone_element_doubler_aux_data;
-        ret->func = func;
-        ret->funcdata = funcdata;
-
-        return (NpyAuxData *)ret;
-    }
+an element copier function as a primitive.
+
+.. code-block:: c
+
+   typedef struct {
+       NpyAuxData base;
+       ElementCopier_Func *func;
+       NpyAuxData *funcdata;
+   } eldoubler_aux_data;
+
+   void free_element_doubler_aux_data(NpyAuxData *data)
+   {
+       eldoubler_aux_data *d = (eldoubler_aux_data *)data;
+       /* Free the memory owned by this auxdata */
+       NPY_AUXDATA_FREE(d->funcdata);
+       PyArray_free(d);
+   }
+
+   NpyAuxData *clone_element_doubler_aux_data(NpyAuxData *data)
+   {
+       eldoubler_aux_data *ret = PyArray_malloc(sizeof(eldoubler_aux_data));
+       if (ret == NULL) {
+           return NULL;
+       }
+
+       /* Raw copy of all data */
+       memcpy(ret, data, sizeof(eldoubler_aux_data));
+
+       /* Fix up the owned auxdata so we have our own copy */
+       ret->funcdata = NPY_AUXDATA_CLONE(ret->funcdata);
+       if (ret->funcdata == NULL) {
+           PyArray_free(ret);
+           return NULL;
+       }
+
+       return (NpyAuxData *)ret;
+   }
+
+   NpyAuxData *create_element_doubler_aux_data(
+                               ElementCopier_Func *func,
+                               NpyAuxData *funcdata)
+   {
+       eldoubler_aux_data *ret = PyArray_malloc(sizeof(eldoubler_aux_data));
+       if (ret == NULL) {
+           PyErr_NoMemory();
+           return NULL;
+       }
+       memset(&ret, 0, sizeof(eldoubler_aux_data));
+       ret->base->free = &free_element_doubler_aux_data;
+       ret->base->clone = &clone_element_doubler_aux_data;
+       ret->func = func;
+       ret->funcdata = funcdata;
+
+       return (NpyAuxData *)ret;
+   }
```

Comment:
Mind splitting this to it's own PR? I assume this works fine on Sphinx 2, and I'm getting fed up of scrolling past it reviewing this diff!

Reply:
(but also, thanks for spotting and improving this!)

Reply:
Done in #17633, along with a bunch more like it.